runtime/major_gc.c

```diff
@@ -607,10 +607,18 @@ static void adopt_orphaned_work (int expected_status)
   while (f != NULL) {
     myf = domain_state->final_info;
     CAMLassert (caml_gc_phase == Phase_sweep_and_mark_main);
-    /* Since we are in [Phase_sweep_and_mark_main], the current domain has not
-       updated its finalisers. */
-    CAMLassert (!myf->updated_first);
-    CAMLassert (!myf->updated_last);
+
+    /* updated_first/last may be true if the current domain is terminating
+       and has orphaned some finalisers but now has to adopt back (the same
+       or other) finalisers. */
+    if (myf->updated_first){
+      (void)caml_atomic_counter_incr(&num_domains_to_final_update_first);
```

Comment:
Reminder to document the updated behaviour here: https://github.com/damiendoligez/ocaml/blob/ce70b370545e4d9357295b917352996026b36dff/runtime/major_gc.c#L81. 

The `num_domains_to_final_update*` counters may increment the counter if there is orphaned work.